qemu/tests/ioeventfd.py

```diff
@@ -0,0 +1,163 @@
+                For linux: dd if=/dev/zero of=/home/test bs=1M count=4096
+                oflag=direct.
+                For windows: We can load some big data to the disk via
+                DummyCMD.exe tools or other tools.
+            1.3 Reboot the guest.
+            1.4 Run iozone in the guest.
+            1.5 Power off the guest.
+            1.6 Repeat the step 1.1-1.5 with ioeventfd=on.
+        Scenario 2:
+            2.1 Boot guest with ioeventfd=off.
+            2.2 Execute info qtree in QMP monitor, info qtree should show the
+                ioeventfd = false.
+            2.3 Check the ioeventfd=off via /proc/$PID/fd/.
+            2.4 Boot guest with ioeventfd=on.
+            2.5 Execute info qtree in QMP monitor, info qtree should show the
+                ioeventfd = true.
+            2.6 Check the ioeventfd=on via /proc/$PID/fd/.
+            2.7 Compare the output of 'ls -l /proc/$PID/fd/', the fds with
+                "off" should be less than the one with "on".
+
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    """
+    def _set_ioeventfd_options():
+        """ Set the ioeventfd options. """
+        if params['drive_format'] == 'virtio':
+            params['blk_extra_params_image1'] = ioeventfd
+            dev_id = 'image1'
+        elif params['drive_format'] == 'scsi-hd':
+            params['bus_extra_params_image1'] = ioeventfd
+            dev_id = 'virtio_scsi_pci0'
+        return dev_id
+
+    def _dd_test():
+        """ Execute dd testing inside guest. """
+        logging.info('Doing dd testing inside guest.')
+        dd_cmd = params['dd_cmd']
+        if os_type == 'windows':
+            dd_cmd = utils_misc.set_winutils_letter(session, dd_cmd)
+        o = session.cmd(dd_cmd, float(params['dd_timeout']))
+        logging.debug(o)
+
+    def _iozone_test(session):
+        """ Execute iozone testing inside guest. """
+        logging.info('Doing iozone inside guest.')
+        if os_type == 'windows':
+            session = utils_test.qemu.windrv_check_running_verifier(
+                session, vm, test, params["driver_name"])
+        try:
+            iozone = generate_instance(params, session, 'iozone')
+            iozone.run(params['iozone_options'], float(params['iozone_timeout']))
+        finally:
+            iozone.clean()
+            return session
+
+    def _check_property(vm, ioeventfd_opt):
+        """
+        Check the value of ioeventfd by sending "info qtree" in QMP.
+        """
+        ioevent_qtree_val = 'true' if 'on' in ioeventfd_opt else 'false'
+        logging.info('Execute info qtree in QMP monitor.')
+        qtree = qemu_qtree.QtreeContainer()
+        qtree.parse_info_qtree(vm.monitor.info('qtree'))
+        for node in qtree.get_nodes():
+            if isinstance(node, qemu_qtree.QtreeDev) and (
+                    node.qtree.get('id', None) == dev_id):
```

Comment:
if here use node.qtree['type'] == device_type is better?  define device_type in cfg file, it can reduce a function _set_ioeventfd_options() and can re-use line 54 and 57.

Reply:
I think that "node.qtree['type'] == device_type" can't match exactly if there are many devices.